### PR TITLE
fix: numerical errors in GPU recombination

### LINKED ISSUE #852: On GPU, TestTreeRecombination is failing
### What's the problem?

On NVIDIA GPU, TestTreeRecombination::test_reduce is failing.

### How can we reproduce the issue?

Run  tests/unit/test_solvers.py::TestTreeRecombination::test_reduce

### Python version

3.12

### Package version

0.3.1

### Operating system

Ubuntu 24.04 with NVIDIA GPU

### Other packages

jax[cuda12]

### Relevant log output

FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-random-implicit-explicit-context0-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-random-implicit-explicit-context0-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-partial-null-implicit-explicit-context0-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-partial-null-implicit-explicit-context0-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-partial-null-implicit-explicit-context0-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-partial-null-implicit-explicit-context0-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-partial-null-explicit-context2-False] - assert 16 <= Array(11, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-partial-null-explicit-context2-True] - assert 16 <= Array(11, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-random-implicit-explicit-context0-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-random-implicit-explicit-context0-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-random-explicit-context2-False] - assert 16 <= Array(11, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-random-explicit-context2-True] - assert 16 <= Array(11, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-null-implicit-explicit-context0-False] - assert 11 <= (11 - (11 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-null-implicit-explicit-context0-True] - assert 11 <= (11 - (11 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-null-explicit-context2-False] - assert 14 <= Array(1, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-null-explicit-context2-True] - assert 14 <= Array(1, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-null-implicit-explicit-context0-False] - assert 11 <= (11 - (11 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-null-implicit-explicit-context0-True] - assert 11 <= (11 - (11 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-full_rank-implicit-explicit-context0-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-full_rank-implicit-explicit-context0-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-full_rank-explicit-context2-False] - assert 6 <= Array(4, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-full_rank-explicit-context2-True] - assert 6 <= Array(4, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-full_rank-implicit-explicit-context0-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-full_rank-implicit-explicit-context0-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-rank_deficient-implicit-explicit-context0-False] - assert 3 <= (4 - (4 - Array(2, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-rank_deficient-implicit-explicit-context0-True] - assert 3 <= (4 - (4 - Array(2, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-rank_deficient-explicit-context2-False] - assert 3 <= Array(2, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-rank_deficient-explicit-context2-True] - assert 3 <= Array(2, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-rank_deficient-implicit-explicit-context0-False] - assert 3 <= (4 - (4 - Array(2, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-rank_deficient-implicit-explicit-context0-True] - assert 3 <= (4 - (4 - Array(2, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-excessive_test_functions-implicit-explicit-context0-False] - assert 84 <= (128 - (128 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-excessive_test_functions-implicit-explicit-context0-True] - assert 84 <= (128 - (128 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-excessive_test_functions-implicit-context1-False] - assert 84 <= (128 - (128 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-excessive_test_functions-implicit-context1-True] - assert 84 <= (128 - (128 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-excessive_test_functions-explicit-context2-False] - assert 84 <= Array(1, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[without_jit-excessive_test_functions-explicit-context2-True] - assert 84 <= Array(1, dtype=int32)
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-excessive_test_functions-implicit-explicit-context0-False] - assert 84 <= (128 - (128 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-excessive_test_functions-implicit-explicit-context0-True] - assert 84 <= (128 - (128 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-excessive_test_functions-implicit-context1-False] - assert 84 <= (128 - (128 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestTreeRecombination::test_reduce[with_jit-excessive_test_functions-implicit-context1-True] - assert 84 <= (128 - (128 - Array(1, dtype=int32)))

### LINKED ISSUE #853: On GPU, TestCaratheodoryRecombination is failing
### What's the problem?

On NVIDIA GPU, TestCaratheodoryRecombination is failing.

### How can we reproduce the issue?

Run  tests/unit/test_solvers.py::TestCaratheodoryRecombination

### Python version

3.12

### Package version

0.3.1

### Operating system

Ubuntu 24.04 with NVIDIA GPU

### Other packages

jax[cuda12]

### Relevant log output

FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[with_jit-random-implicit-explicit-context0-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[with_jit-random-implicit-explicit-context0-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[with_jit-partial-null-implicit-explicit-context0-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[with_jit-partial-null-implicit-explicit-context0-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[with_jit-partial-null-implicit-context1-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[with_jit-partial-null-implicit-context1-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-partial-null-implicit-explicit-context0-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-partial-null-implicit-explicit-context0-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-partial-null-implicit-context1-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-partial-null-implicit-context1-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-partial-null-explicit-context2-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-partial-null-explicit-context2-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-random-implicit-explicit-context0-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-random-implicit-explicit-context0-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-random-explicit-context2-False] - assert 25 <= Array(11, dtype=int32)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-random-explicit-context2-True] - assert 25 <= Array(11, dtype=int32)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-null-implicit-explicit-context0-False] - assert 11 <= (11 - (11 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-null-implicit-explicit-context0-True] - assert 11 <= (11 - (11 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-null-explicit-context2-False] - assert 23 <= Array(1, dtype=int32)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-null-explicit-context2-True] - assert 23 <= Array(1, dtype=int32)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[with_jit-null-implicit-explicit-context0-False] - assert 11 <= (11 - (11 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[with_jit-null-implicit-explicit-context0-True] - assert 11 <= (11 - (11 - Array(1, dtype=int32)))
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-full_rank-implicit-explicit-context0-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-full_rank-implicit-explicit-context0-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-full_rank-explicit-context2-False] - assert 18 <= Array(4, dtype=int32)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-full_rank-explicit-context2-True] - assert 18 <= Array(4, dtype=int32)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[with_jit-full_rank-implicit-explicit-context0-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[with_jit-full_rank-implicit-explicit-context0-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-rank_deficient-implicit-explicit-context0-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-rank_deficient-implicit-explicit-context0-True] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-rank_deficient-explicit-context2-False] - assert 26 <= Array(2, dtype=int32)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[without_jit-rank_deficient-explicit-context2-True] - assert 26 <= Array(2, dtype=int32)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[with_jit-rank_deficient-implicit-explicit-context0-False] - assert Array(False, dtype=bool)
FAILED tests/unit/test_solvers.py::TestCaratheodoryRecombination::test_reduce[with_jit-rank_deficient-implicit-explicit-context0-True] - assert Array(False, dtype=bool)

### PR DESCRIPTION
**Pull request by rg936672**
### PR Type

- Bugfix

### Description

Closes #852. Closes #853. Fixes an incorrect update applied in the elimination step of the recombination solvers.

### How Has This Been Tested?

Existing tests pass as expected, even on a GPU machine.

### Does this PR introduce a breaking change?

No.

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have ensured my code is easy to understand, including docstrings and comments where necessary.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I have updated CHANGELOG.md, if appropriate.

### CODE REVIEW

**Comment by tp832944** [comment]
> Should this PR also resolve #852? Additionally, would it be possible to add a GPU runner to the CI tests so we can avoid these issues going silently unnoticed in the future (will need to open a new issue for this)?

We had thought about this previously, but decided against it primarily on cost grounds. We do run GPU tests manually just before each release. A few older GPU bugs may have gone unnoticed until now because of a configuration issue with the GPU.

**Review by tc85324 — CHANGES_REQUESTED** [review]
I think your idea is good, and a valid approach! However, I believe we can achieve the same thing (fixing both #853 and #854) by simply correcting the following line, which is currently incorrect:
```python
# Old (incorrect)
updated_null_space_basis.at  = updated_null_space_basis.at[basis_index].set(0)
# New (correct)
updated_null_space_basis.at  = updated_null_space_basis.at[:, elimination_index].set(0)
```
In my own local tests, this appears to work on CPU and GPU devices; all tests pass on CPU and GPU devices.

The only other thing that might be worthwhile is to indicate to users (via the docstring or a warning) that enabling double precision ("jax_enable_x64") is sometimes required to obtain correct results.

**Comment by rg936672** [comment]
Ah, that does make sense - I'd just assumed the code was correct as I wasn't super familiar with the algorithm!

File: CHANGELOG.md [hunk 1]
```diff
@@ -18,6 +18,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 
 - `MMD.compute` no longer returns `nan`. (https://github.com/gchq/coreax/issues/855)
+- CaratheodoryRecombination now generates correct coresets on GPU machines
```
Comment (tc85324) [inline CHANGELOG.md:21]: Should this now be updated to mention the fix for all the recombination solvers, and should there be links to the underlying issues (in addition to the existing PR link)? 
Maybe something like:
```
Corrected an implementation error in `coreax.solvers.CaratheodoryRecombination`
which caused numerical instability when using either `CaratheodoryRecombination`
or `TreeRecombination` on GPU machines. 
(#852, #853, #874)
```

**Review by tc85324 — CHANGES_REQUESTED** [review]
Very small comment on updating the CHANGELOG, but otherwise looks good to me!

**Review by tc85324 — APPROVED** [review]

**Pull request merged**